WIP: Basic algorithm for printing floats with %f

JohnnyDingYQ opened the pull request:
[Issue #74](#74)
This is as WIP pull request that aims to enable string formatting of strings. I implemented the "basic conversion algorithm" outlined in the [Ryu paper](https://dl.acm.org/doi/pdf/10.1145/3192366.3192369). A future port to Ryu, the improved algorithm, can be planned in the future once the existing code is stable

The major additions are:
- Created [lib/util/Floats.v3](lib/util/Floats.v3) for core floating rendering code
- Modified [test/lib/StringBuilderTest.v3](test/lib/StringBuilderTest.v3) for %f tests

JohnnyDingYQ commented on lib/util/StringBuilder.v3:
```diff
@@ -351,6 +351,11 @@ class StringBuilder {
 	}
 
+	def putf(f: float) -> this {
+		acquire(16);
```
Yes, my new guess is 41, which is len(float_max) + sign + dot = 39 + 1 + 1 = 41. I can validate this by adding more tests.

titzer commented on lib/util/Floats.v3:
```diff
@@ -0,0 +1,665 @@
+// Copyright 2024 Virgil authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+// ====================== Start of Ryu license ========================
```
Since both Virgil and Ulf used the Apache 2.0 license, you can combine these notices, e.g.

```
// Copyright 2025 Virgil authors.
// Copyright 2018 Ulf Adams (see below).

...

// Based on code by Ulf Adams (further reference)
```

titzer commented on lib/util/Floats.v3:
```diff
@@ -0,0 +1,665 @@
+			a[pos] = 'N';
+			return 3;
+		}
+		if (exponent != 0) {
```
You could use `Arrays.copyInto` for these sequences.

Also side note: I was planning on refactoring both `Ints` and `Longs` to use `Range<byte>` but until then, it's good that you're consistent with them.

titzer commented on lib/util/StringBuilder.v3:
```diff
@@ -351,6 +351,11 @@ class StringBuilder {
 		return (-1, offset);
 	}
 
+	def putf(f: float) -> this {
```
Thanks for adding the case, though I've reconsidered and we should add that one line in a following PR to avoid including the code by default (DCE won't necessarily be smart enough to remove it just yet).

titzer commented on lib/util/StringBuilder.v3:
```diff
@@ -41,13 +41,14 @@ class StringBuilder {
 			x: i32 => putd_i32(x);
 			x: u64 => putd_u64(x);
 			x: i64 => putd_i64(x);
+			x: float => putf(x);
```
A line for a future PR.

titzer reviewed (commented):
Looks good with some minor comments.

Thanks again!

JohnnyDingYQ commented:
Hi, I have moved the tests into a new file `FloatsTest.v3` and completed the function that would permit the `%.Nd` feature. The relevant tests are presents and passing, so I think this PR is good to go. I reverted changes in `StringBuilder.v3` and `StringBuilderTest.v3` and kept all modifications in `Floats.v3` and `FloatsTest.v3` (new files).

titzer commented:
> Hi, I have moved the tests into a new file `FloatsTest.v3` and completed the function that would permit the `%.Nd` feature. The relevant tests are presents and passing, so I think this PR is good to go. I reverted changes in `StringBuilder.v3` and `StringBuilderTest.v3` and kept all modifications in `Floats.v3` and `FloatsTest.v3` (new files).

Thanks!
